### PR TITLE
Add the ability to return the raw detail of the api problem

### PR DESCRIPTION
**Pull request by MatthiasKuehneEllerhold**
In my custom error logger I want to log the raw detail of an api problem (e. g. an exception) in order to have the stacktrace in the same format as any other exception thrown in a ZF2 application.

This PR alters the `ApiProblem::getDetail()` method:
* makes it public to allow retrieving the detail
* adds a parameter "bool $raw" (default false) to allow the return of the raw detail (the exception) instead of an string.
* also makes the getStatus() method public too

Or would you rather prefer a new method `ApiProblem::getRawDetail()` instead of altering the existing method?

### CODE REVIEW

**Comment by weierophinney** [comment]
I would prefer a different method, for two reasons:

- Changing the signature of a class breaks existing extensions.
- Using flags to change the behavior of a method generally makes documenting and usage more difficult.

**Comment by MatthiasKuehneEllerhold** [comment]
@weierophinney I've added `getRawDetail` like you requested. Any thoughts?

File: src/ApiProblem.php [hunk 1]
```diff
@@ -250,14 +250,24 @@ protected function getDetail()
     }
 
     /**
```
Comment (TomHAnderson) [inline src/ApiProblem.php]: Throwing my 2 cents in here, I don't like the term 'raw' in my applications.  I'm in favor of finding a better function name.

Comment (weierophinney) [inline src/ApiProblem.php]: We've used the term "raw" elsewhere in the framework, including zend-inputfilter; this would at least be consistent.

**Comment by MatthiasKuehneEllerhold** [comment]
Rebased on current master an added some unit tests.

**Comment by MatthiasKuehneEllerhold** [comment]
Travis failed with this error:
```
$ phpenv global 7.1
rbenv: version `7.1' not installed
The command "phpenv global 7.1" failed and exited with 1 during .
```

I'm guessing my unit tests aren't responsible?

**Comment by MatthiasKuehneEllerhold** [comment]
Any news?

File: src/ApiProblem.php [hunk 2]
```diff
@@ -239,7 +239,7 @@ public function setDetailIncludesStackTrace($flag)
      *
      * @return string
      */
-    protected function getDetail()
+    public function getDetail()
```
Comment (weierophinney) [inline src/ApiProblem.php]: The visibility on this should not change; access to processed detail is available via direct property access already (`$problem->detail`). (I'll revert on merge)

File: src/ApiProblem.php [hunk 3]
```diff
@@ -256,7 +266,7 @@ protected function getDetail()
      *
      * @return string
      */
-    protected function getStatus()
+    public function getStatus()
```
Comment (weierophinney) [inline src/ApiProblem.php]: Same comment here: access is available via direct property access (`$problem->status`). I'll revert on merge.

File: test/ApiProblemTest.php [hunk 4]
```diff
@@ -54,6 +54,8 @@ public function testErrorAsDetails()
         $this->assertEquals(705, $payload['status']);
         $this->assertArrayHasKey('detail', $payload);
         $this->assertEquals('error message', $payload['detail']);
+        $this->assertEquals('error message', $apiProblem->getDetail());
```
Comment (weierophinney) [inline test/ApiProblemTest.php]: `s/$apiProblem->getDetail()/$apiProblem->detail/` (I'll do this on merge)

File: test/ApiProblemTest.php [hunk 5]
```diff
@@ -71,6 +73,8 @@ public function testDetailStringIsUsedVerbatim()
         $payload = $apiProblem->toArray();
         $this->assertArrayHasKey('detail', $payload);
         $this->assertEquals('foo', $payload['detail']);
+        $this->assertEquals('foo', $apiProblem->getDetail());
```
Comment (weierophinney) [inline test/ApiProblemTest.php]: Same comment here.

File: test/ApiProblemTest.php [hunk 6]
```diff
@@ -80,6 +84,8 @@ public function testExceptionMessageIsUsedForDetail()
         $payload = $apiProblem->toArray();
         $this->assertArrayHasKey('detail', $payload);
         $this->assertEquals($exception->getMessage(), $payload['detail']);
+        $this->assertEquals('exception message', $apiProblem->getDetail());
```
Comment (weierophinney) [inline test/ApiProblemTest.php]: And finally, here.

**Comment by MatthiasKuehneEllerhold** [comment]
You're right, didnt see the magic `__get` method. This would mean that `$problem->detail` is kinda public too? Seems like this whole PR is unnecessary then? Sorry about that!

**Comment by weierophinney** [comment]
> This would mean that $problem->detail is kinda public too? Seems like this whole PR is unnecessary then?

Now that I look at it closer... yes, you're right. I was thinking property access was calling the internal methods (`getDetail()`, `getStatus()`, etc.), which would mean you'd need the additional accessor if you wanted the raw values, as `getDetail()` varies behavior based on the value of `$detail`. However, in reviewing, it gives direct access to the properties, which means that calling `$apiProblem->detail` gives you the raw access you want.

So... closing!

**Pull request closed without merging**